### PR TITLE
[fix][client]Fix client memory limit currentUsage leak and semaphore release duplicated in ProducerImpl

### PR DESCRIPTION
**Pull request by Nicklee007**
### Motivation

Fix client memory limit `currentUsage` leak in `ProducerImpl`. When our pulsar cluster occur some error,  producer send message fail and we find the `currentUsage` always  keep high value like the leaked, and cause the producer send rate is slow.
And find producer semaphore release duplicated when `createOpSendMsg`  occur some excrption.

Follow 1 point only release the message count semaphore, but not release the memory limit.
**memory limit currentUsage leak point**
https://github.com/apache/pulsar/blob/c217b8f559292fd34c6a4fb4b30aab213720d962/pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java#L2031-L2033

**producer semaphore release duplicated**
https://github.com/apache/pulsar/blob/4d64e2e66689381ebbb94fbfc03eb4e1dfba0405/pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java#L2116-L2120

```
After the exception the  memory limit leak occured.
org.apache.pulsar.client.api.PulsarClientException$TimeoutException: The producer XXXX-366-15151 can not send message to the topic persistent://XXXX/XXXX/XXXX within given timeout : createdAt 30.005 seconds ago, firstSentAt 30.005 seconds ago, lastSentAt 30.005 seconds ago, retryCount 1
        at org.apache.pulsar.client.impl.ProducerImpl$OpSendMsg.sendComplete(ProducerImpl.java:1287)
        at org.apache.pulsar.client.impl.ProducerImpl.lambda$failPendingMessages$18(ProducerImpl.java:1826)
        at java.base/java.util.ArrayDeque.forEach(ArrayDeque.java:889)
        at org.apache.pulsar.client.impl.ProducerImpl$OpSendMsgQueue.forEach(ProducerImpl.java:1369)
        at org.apache.pulsar.client.impl.ProducerImpl.failPendingMessages(ProducerImpl.java:1816)
        at org.apache.pulsar.client.impl.ProducerImpl.lambda$failPendingMessages$19(ProducerImpl.java:1848)
        at org.apache.pulsar.shade.io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
        at org.apache.pulsar.shade.io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:469)
        at org.apache.pulsar.shade.io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:384)
        at org.apache.pulsar.shade.io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986)
        at org.apache.pulsar.shade.io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at org.apache.pulsar.shade.io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:834)

```

### Modifications

1. add the `MemoryLimitController` release.

### Documentation

- [X] `doc-not-needed` 
(Please explain why)

### CODE REVIEW

**Comment by codelipenghui** [comment]
A related one https://github.com/apache/pulsar/pull/16835

**Comment by codelipenghui** [comment]
@Nicklee007 Could you please help add some unit tests?

**Comment by Shoothzj** [comment]
+1 we need some tests to cover this change

**Comment by Nicklee007** [comment]
> @Nicklee007 Could you please help add some unit tests?

@codelipenghui @Shoothzj  Add some unit test, PTAL Thx~

File: pulsar-broker/src/test/java/org/apache/pulsar/client/impl/ProducerMemoryLimitTest.java [hunk 1]
```diff
@@ -69,6 +69,35 @@ public void testProducerTimeoutMemoryRelease() throws Exception {
...
+                .maxPendingMessages(0)
+                .enableBatching(true)
+                .batchingMaxPublishDelay(100, TimeUnit.MILLISECONDS)
+                .batchingMaxBytes(12)
+                .create();
+        this.stopBroker();
+        try {
+            producer.newMessage().value("memory-test".getBytes(StandardCharsets.UTF_8)).sendAsync();
+            try {
+                producer.newMessage().value("memory-test".getBytes(StandardCharsets.UTF_8)).sendAsync().get();
```
Comment (Jason918) [inline pulsar-broker/src/test/java/org/apache/pulsar/client/impl/ProducerMemoryLimitTest.java:88]: This seems not be able cover the case in `batchMessageAndSend`

Comment (Nicklee007) [reply pulsar-broker/src/test/java/org/apache/pulsar/client/impl/ProducerMemoryLimitTest.java:88]: @Jason918 Through some test, find the memory limit release operator add in `batchMessageAndSend` will cause duplicate release with add in `failPendingBatchMessages`.  So  remove the fixed in `batchMessageAndSend`.  
Please take another look, Thks.

**Comment by Nicklee007** [comment]
@codelipenghui @Jason918 @Shoothzj  Also find producer semaphore release duplicated in `batchMessageAndSend`, if some error cause in  `batchMessageAndSend`, the `semaphoreRelease` will be invoked in `failPendingBatchMessages` .
Fixed it and add some unit test to cover it.

**Review by gaozhangmin — APPROVED** [review]

**Review by Shoothzj — APPROVED** [review]
Nice catch

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java [hunk 2]
```diff
@@ -2113,10 +2115,8 @@ private void batchMessageAndSend(boolean shouldScheduleNextBatchFlush) {
                 for (OpSendMsg opSendMsg : opSendMsgs) {
                     processOpSendMsg(opSendMsg);
                 }
-            } catch (PulsarClientException e) {
-                semaphoreRelease(batchMessageContainer.getNumMessagesInBatch());
             } catch (Throwable t) {
-                semaphoreRelease(batchMessageContainer.getNumMessagesInBatch());
+                // Nothing to do.
```
Comment (Technoboy-) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java:2119]: We can remove this comment...

Comment (Nicklee007) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java]: Removed, PTAL.

**Comment by HQebupt** [comment]
Nice catch！！！

**Review by HQebupt — APPROVED** [review]

**Review by Technoboy- — APPROVED** [review]

**Review by mattisonchao — APPROVED** [review]

**Comment by mattisonchao** [comment]
@codelipenghui  Please review it again. thanks!

**Review by shibd — APPROVED** [review]

**Review by Jason918 — APPROVED** [review]
LGTM

**Comment by mattisonchao** [comment]
@Nicklee007  Could you help to open another PR to branch-2.9?

**Comment by BewareMyPower** [comment]
Remove the `release/2.7.5` label because memory limiter was introduced since 2.8, see https://github.com/apache/pulsar/pull/8965.

**Comment by Nicklee007** [comment]
> @Nicklee007 Could you help to open another PR to branch-2.9?

@mattisonchao ok, I'll cherry-pick it to branch 2.9.

**Comment by BewareMyPower** [comment]
Move this PR to the next release of 2.8.5. For the release manager of next 2.8.x release, this PR requires an extra import of `Mockito` in https://github.com/apache/pulsar/pull/16835. However, even after fixing the compilation error, the test still failed. So let's just delay this PR to the next release.

@Nicklee007 If you want to include this PR in 2.8.4, please open a new PR to migrate this PR to branch-2.8.

**Comment by Nicklee007** [comment]
> @Nicklee007 Could you help to open another PR to branch-2.9?

@mattisonchao cherry-pick to PR #16971  in branch 2.9.

**Comment by Nicklee007** [comment]
> Remove the `release/2.7.5` label because memory limiter was introduced since 2.8, see #8965.

@BewareMyPower @Jason918 Created a new PR #16972 to branch-2.7 which only fix the semaphore release duplicated problem in ProducerImpl, PTAL.

**Comment by Nicklee007** [comment]
> Move this PR to the next release of 2.8.5. For the release manager of next 2.8.x release, this PR requires an extra import of `Mockito` in #16835. However, even after fixing the compilation error, the test still failed. So let's just delay this PR to the next release.
> 
> @Nicklee007 If you want to include this PR in 2.8.4, please open a new PR to migrate this PR to branch-2.8.

@BewareMyPower Created a new PR #16985 branch-2.8, PTAL.

**Pull request merged**